File: android/notion/src/main/java/com/prince/notion/NotionIntegration.java
package com.prince.notion;

import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Looper;
import android.view.inputmethod.EditorInfo;

import androidx.annotation.Nullable;

import com.prince.notion.ui.NotionConnectActivity;
import com.prince.kbd.core.CommandSpec;
import com.prince.kbd.core.IntegrationContext;
import com.prince.kbd.core.IntegrationSession;
import com.prince.kbd.core.KeyValueStore;
import com.prince.kbd.core.KeyboardIntegration;
import com.prince.kbd.core.GeminiService;

import org.json.JSONArray;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Notion integration. Contributes {@code /notion} (and {@code /note} alias): structures
 * the prompt with the LLM, creates a Notion page under the user's chosen parent, and
 * surfaces the result via system notification.
 */
public final class NotionIntegration implements KeyboardIntegration {

    private static final Set<String> AFFINITY = new HashSet<>(Arrays.asList(
            "com.Slack",
            "com.whatsapp",
            "com.notion.id",
            "com.google.android.gm",
            "org.telegram.messenger",
            "com.discord",
            "com.android.chrome"));

    @Override public String id() { return "notion"; }

    @Override
    @Nullable
    public IntegrationSession activate(EditorInfo info, IntegrationContext ctx) {
        // Slash-driven, not contextual — no per-input chip.
        return null;
    }

    @Override
    public List<CommandSpec> commands() {
        return Arrays.asList(
                new CommandSpec("notion", "Notion page", "📓", true, this::handleNotion, AFFINITY),
                new CommandSpec("note",   "Notion page", "📓", true, this::handleNotion, AFFINITY));
    }

    private void handleNotion(String prompt, IntegrationContext ctx) {
        KeyValueStore store = ctx.store("notion");
        NotionAuth auth = new NotionAuth(store);
        if (!auth.isSignedIn()) {
            ctx.showBanner("Connect Notion in the Turtle app", 1800L);
            ctx.openScreen("notion-connect");
            return;
        }
        String parent = store.getString(NotionKeys.DEFAULT_PARENT, "");
        if (parent == null || parent.isEmpty()) {
            ctx.showBanner("Pick a Notion parent page in the Turtle app", 1800L);
            ctx.openScreen("notion-connect");
            return;
        }
        if (prompt == null || prompt.trim().isEmpty()) {
            ctx.showBanner("Type something after /notion", 1500L);
            return;
        }

        ctx.showBanner("📓 Creating Notion page…", 1200L);

        final Context appContext = ctx.appContext();
        final String token = auth.accessToken();
        final String userPrompt = prompt;
        final GeminiService llm = ctx.ai();

        NotionLlmBridge.structure(userPrompt, llm, new NotionLlmBridge.Callback() {
            @Override public void onStructured(String title, JSONArray blocks) {
                new NotionClient(token).createPage(parent, title, blocks, new NotionClient.PageCallback() {
                    @Override public void onSuccess(String pageId, String pageUrl) {
                        post(() -> NotionResultNotifier.notifySuccess(
                                appContext, title, pageUrl == null ? canonicalUrl(pageId) : pageUrl));
                    }
                    @Override public void onError(String reason) {
                        post(() -> NotionResultNotifier.notifyError(appContext, userPrompt, reason));
                    }
                });
            }
            @Override public void onError(String reason) {
                post(() -> NotionResultNotifier.notifyError(appContext, userPrompt, "LLM: " + reason));
            }
        });
    }

    private static String canonicalUrl(String pageId) {
        // Notion web URLs use the dash-stripped id.
        if (pageId == null) return "https://www.notion.so";
        return "https://www.notion.so/" + pageId.replace("-", "");
    }

    private static void post(Runnable r) {
        new Handler(Looper.getMainLooper()).post(r);
    }
}

File: ios/TurtleKeyboardExtension/Integration/Notion/NotionIntegration.swift
import Foundation
#if os(iOS)
import UIKit

/// Commands-only Notion integration. Contributes `/notion <prompt>` and
/// `/note <prompt>` (alias) — both create a Notion page under the user's
/// chosen parent. Mirrors Android's `NotionIntegration`.
///
/// Behavior is fire-and-forget:
///   1. User taps Send on `/notion <prompt>`.
///   2. Banner flashes "Creating Notion page…".
///   3. Background: LLM structures the prompt → Notion API creates the page.
///   4. Banner flashes the result (success copies the page URL to clipboard).
///
/// If the user hasn't completed OAuth + parent picking, the handler shows
/// a banner pointing them to the host app's "Connect Notion" screen.
final class NotionIntegration: KeyboardIntegration {

    let id = "notion"

    func commands() -> [CommandSpec] {
        [
            CommandSpec(
                name: "notion", label: "Notion page", emoji: "📓", needsPrompt: true,
                handler: { prompt, ctx in Self.handle(prompt: prompt, ctx: ctx) }
            ),
            CommandSpec(
                name: "note", label: "Notion page", emoji: "📓", needsPrompt: true,
                handler: { prompt, ctx in Self.handle(prompt: prompt, ctx: ctx) }
            ),
        ]
    }

    static func handle(prompt: String, ctx: IntegrationContext) {
        let store = ctx.store
        let token = store.string(forKey: NotionKeys.accessToken, fallback: "")
        guard !token.isEmpty else {
            ctx.showBanner("Connect Notion in the Turtle app", autoHideMs: 1800)
            ctx.openScreen("notion-connect")
            return
        }
        let parent = store.string(forKey: NotionKeys.defaultParent, fallback: "")
        guard !parent.isEmpty else {
            ctx.showBanner("Pick a Notion parent page in the Turtle app", autoHideMs: 1800)
            ctx.openScreen("notion-connect")
            return
        }
        let trimmed = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            ctx.showBanner("Type something after /notion", autoHideMs: 1500)
            return
        }

        ctx.showBusy("📓 Creating Notion page…")
        let llm = ctx.llm

        NotionLlmBridge.structure(userPrompt: trimmed, llm: llm) { result in
            switch result {
            case .failure(let error):
                ctx.showBanner("⚠️ Notion: \(error.localizedDescription)", autoHideMs: 2500)
            case .success(let parsed):
                Task {
                    do {
                        let (pageId, pageURL) = try await NotionClient.createPage(
                            accessToken: token,
                            parentPageId: parent,
                            title: parsed.title,
                            blocks: parsed.blocks
                        )
                        let url = pageURL ?? canonicalURL(pageId: pageId)
                        await MainActor.run {
                            UIPasteboard.general.string = url
                            ctx.showBanner("📓 Page created — link copied", autoHideMs: 2000)
                        }
                    } catch {
                        await MainActor.run {
                            ctx.showBanner("⚠️ Notion: \(error.localizedDescription)", autoHideMs: 2500)
                        }
                    }
                }
            }
        }
    }

    private static func canonicalURL(pageId: String) -> String {
        // Notion's web URL strips dashes from the id.
        let stripped = pageId.replacingOccurrences(of: "-", with: "")
        return "https://www.notion.so/\(stripped)"
    }
}
#endif
